File: Sources/NotchAgent/Core/Services/RefreshScheduler.swift
import AppKit
import Foundation

/// Coalesces repeated refresh clicks without dropping a forced refresh that
/// arrives while the periodic loop is still running.
struct RefreshRequestQueue {
    private(set) var isRunning = false
    private var forcedRefreshPending = false

    mutating func begin(force: Bool) -> Bool {
        guard !isRunning else {
            if force { forcedRefreshPending = true }
            return false
        }
        isRunning = true
        if force { forcedRefreshPending = false }
        return true
    }

    /// Returns true when one forced refresh must run immediately afterward.
    mutating func finish() -> Bool {
        isRunning = false
        return forcedRefreshPending
    }
}

struct RefreshGenerationTracker {
    private(set) var counter: UInt64 = 0
    private var providerGeneration: [ProviderID: UInt64] = [:]
    private var accountGeneration: [UUID: UInt64] = [:]

    mutating func beginProvider(_ provider: ProviderID) -> UInt64 {
        counter &+= 1
        providerGeneration[provider] = counter
        return counter
    }

    mutating func beginAccount(_ accountID: UUID) -> UInt64 {
        counter &+= 1
        accountGeneration[accountID] = counter
        return counter
    }

    func acceptsProvider(_ provider: ProviderID, generation: UInt64) -> Bool {
        providerGeneration[provider] == generation
    }

    func acceptsAccount(_ accountID: UUID, generation: UInt64) -> Bool {
        accountGeneration[accountID] == generation
    }

    func accountIsNewer(_ accountID: UUID, than generation: UInt64) -> Bool {
        (accountGeneration[accountID] ?? 0) > generation
    }

    mutating func didApplyProvider(
        _ provider: ProviderID,
        generation: UInt64,
        accountIDs: [UUID] = []
    ) {
        guard acceptsProvider(provider, generation: generation) else { return }
        for accountID in accountIDs {
            if !accountIsNewer(accountID, than: generation) {
                accountGeneration[accountID] = generation
            }
        }
    }
}

/// Central refresh loop: fans out to all providers concurrently, feeds the
/// store, persists snapshots/history. One scheduler owns all polling — no
/// provider ever refreshes itself.
@MainActor
final class RefreshScheduler {
    private let providers: [any UsageProvider]
    private let store: UsageStore
    private let snapshotStore: SnapshotStore
    private let historyStore: HistoryStore
    private let statusPage = StatusPageService()
    private var loopTask: Task<Void, Never>?
    private var refreshQueue = RefreshRequestQueue()
    private var generations = RefreshGenerationTracker()

    init(
        providers: [any UsageProvider],
        store: UsageStore,
        snapshotStore: SnapshotStore,
        historyStore: HistoryStore
    ) {
        self.providers = providers
        self.store = store
        self.snapshotStore = snapshotStore
        self.historyStore = historyStore
    }

    func start() {
        guard loopTask == nil else { return }
        Log.refresh.info("scheduler started")

        NSWorkspace.shared.notificationCenter.addObserver(
            self,
            selector: #selector(systemDidWake),
            name: NSWorkspace.didWakeNotification,
            object: nil
        )

        loopTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                await self.tick()
                let interval = max(15, self.store.settings.refreshIntervalSeconds)
                try? await Task.sleep(for: .seconds(interval))
            }
        }
    }

    func stop() {
        loopTask?.cancel()
        loopTask = nil
        NSWorkspace.shared.notificationCenter.removeObserver(self)
    }

    func refreshNow() {
        Task { await tick(force: true) }
    }

    func refreshAPIAccount(_ accountID: UUID) {
        guard let account = store.settings.apiAccounts.first(where: {
            $0.id == accountID && $0.enabled
        }), let provider = providers.first(where: {
            $0.id == .apiAccounts
        }) as? APIAccountProvider else { return }

        let generation = generations.beginAccount(accountID)
        store.markAccountRefreshing(accountID)
        Task { [weak self] in
            guard let self else { return }
            let usage = await provider.fetchAccountUsage(for: account)
            guard self.generations.acceptsAccount(accountID, generation: generation) else {
                Log.refresh.debug("discarded superseded account refresh")
                return
            }

            var snapshot = self.store.snapshots[.apiAccounts]
                ?? UsageSnapshot(provider: .apiAccounts, health: .noData)
            var usages = snapshot.accountUsage ?? []
            if let index = usages.firstIndex(where: { $0.accountID == accountID }) {
                usages[index] = usage
            } else {
                usages.append(usage)
            }
            snapshot.accountUsage = usages
            snapshot.capturedAt = Date()
            snapshot.health = usages.allSatisfy {
                $0.readStatus == .updated || $0.readStatus == .partial
            } ? .ok : .degraded
            snapshot.note = usages.map(\.summary).joined(separator: " · ")
            self.store.apply(snapshot, updatingAccountIDs: [accountID])
            await self.historyStore.record(snapshot)
            await self.snapshotStore.save(self.store.snapshots)
            await self.historyStore.flush()
        }
    }

    /// Stop + start: applies a changed refresh interval immediately instead of
    /// waiting out the previous (possibly 5-minute) sleep.
    func restart() {
        stop()
        start()
    }

    @objc private func systemDidWake() {
        // Respect a user-initiated pause — waking the lid must not spend probes.
        guard !store.isPaused else { return }
        Log.refresh.info("system woke — refreshing")
        refreshNow()
    }

    private func tick(force: Bool = false) async {
        if store.isPaused && !force {
            return
        }
        // A wake-triggered forced tick may race the periodic loop — never run
        // two ticks interleaved. Forced clicks are queued instead of dropped.
        guard refreshQueue.begin(force: force) else { return }
        let settings = store.settings
        for provider in providers {
            store.markRefreshing(provider.id)
        }

        let results = await withTaskGroup(
            of: (ProviderID, UInt64, Result<UsageSnapshot, Error>).self,
            returning: [(ProviderID, UInt64, Result<UsageSnapshot, Error>)].self
        ) { group in
            for provider in providers {
                let generation = generations.beginProvider(provider.id)
                group.addTask {
                    do {
                        if force {
                            await provider.invalidateCache()
                        }
                        return (
                            provider.id,
                            generation,
                            .success(try await provider.fetchSnapshot(settings: settings))
                        )
                    } catch {
                        return (provider.id, generation, .failure(error))
                    }
                }
            }
            var collected: [(ProviderID, UInt64, Result<UsageSnapshot, Error>)] = []
            for await result in group {
                collected.append(result)
            }
            return collected
        }

        for (providerID, generation, result) in results {
            guard generations.acceptsProvider(providerID, generation: generation) else {
                Log.refresh.debug("discarded superseded provider refresh")
                continue
            }
            switch result {
            case .success(var snapshot):
                var updatedAccountIDs: Set<UUID>?
                if providerID == .apiAccounts {
                    let current = Dictionary(
                        uniqueKeysWithValues: (store.snapshots[.apiAccounts]?.accountUsage ?? [])
                            .map { ($0.accountID, $0) }
                    )
                    var accepted = Set<UUID>()
                    snapshot.accountUsage = (snapshot.accountUsage ?? []).compactMap { usage in
                        guard generations.accountIsNewer(usage.accountID, than: generation) else {
                            accepted.insert(usage.accountID)
                            return usage
                        }
                        return current[usage.accountID]
                    }
                    updatedAccountIDs = accepted
                }
                store.apply(snapshot, updatingAccountIDs: updatedAccountIDs)
                generations.didApplyProvider(
                    providerID,
                    generation: generation,
                    accountIDs: Array(updatedAccountIDs ?? Set(snapshot.accountUsage?.map(\.accountID) ?? []))
                )
                await historyStore.record(snapshot)
                Log.refresh.debug("refreshed \(providerID.rawValue, privacy: .public): \(snapshot.health.rawValue, privacy: .public)")
            case .failure(let error):
                let accountIDs = providerID == .apiAccounts
                    ? Set(store.settings.apiAccounts.filter {
                        $0.enabled && !generations.accountIsNewer($0.id, than: generation)
                    }.map(\.id))
                    : nil
                store.applyFailure(
                    providerID,
                    error: error.localizedDescription,
                    updatingAccountIDs: accountIDs
                )
                Log.refresh.error("refresh failed \(providerID.rawValue, privacy: .public): \(error.localizedDescription, privacy: .public)")
            }
        }

        await snapshotStore.save(store.snapshots)

        // Insights da sessão ativa: um payload por provider, computado uma vez
        // por tick — as views só renderizam (nunca recalculam).
        let payloads = await SessionInsightsEngine.refresh(
            snapshots: store.snapshots,
            dataProvider: LiveSessionDataProvider()
        )
        store.setPayloads(payloads)

        await historyStore.flush()
        await updateSparklines()
        store.setIncident(await statusPage.activeIncident())

        let runQueuedRefresh = refreshQueue.finish()
        if runQueuedRefresh {
            await tick(force: true)
        }
    }

    private func updateSparklines() async {
        for provider in providers {
            let points = await historyStore.series(for: provider.id, lastHours: 24)
            store.updateSparkline(provider.id, values: points.map { Double($0.sessionTokens) })
        }
    }
}

File: windows/NotchAgent.Windows/Services/RefreshScheduler.cs
using Avalonia.Threading;
using NotchAgent.Windows.Models;
using NotchAgent.Windows.Providers;

namespace NotchAgent.Windows.Services;

/// Central refresh loop: fans out to all providers concurrently, feeds the
/// store, persists snapshots. One scheduler owns all polling — no provider
/// ever refreshes itself.
public sealed class RefreshScheduler
{
    private readonly List<IUsageProvider> _providers;
    private readonly UsageStore _store;
    private readonly SnapshotStore _snapshotStore;
    private CancellationTokenSource? _loopCts;
    private bool _tickInFlight;

    public RefreshScheduler(List<IUsageProvider> providers, UsageStore store, SnapshotStore snapshotStore)
    {
        _providers = providers;
        _store = store;
        _snapshotStore = snapshotStore;
    }

    public void Start()
    {
        if (_loopCts is not null) return;
        Log.Refresh.LogInformation("scheduler started");
        var cts = new CancellationTokenSource();
        _loopCts = cts;
        _ = Task.Run(async () =>
        {
            while (!cts.IsCancellationRequested)
            {
                try { await Tick(cts.Token); }
                catch (Exception ex)
                {
                    // Defense in depth: Tick() already catches internally, but
                    // the loop itself must survive anything that slips through.
                    Log.Refresh.LogError("scheduler loop caught: {0}", ex);
                }
                var interval = Math.Max(15, _store.Settings.RefreshIntervalSeconds);
                try { await Task.Delay(TimeSpan.FromSeconds(interval), cts.Token); }
                catch (TaskCanceledException) { return; }
            }
        }, cts.Token);
    }

    public void Stop()
    {
        _loopCts?.Cancel();
        _loopCts = null;
    }

    /// Stop + start: applies a changed refresh interval immediately instead
    /// of waiting out the previous (possibly long) sleep.
    public void Restart()
    {
        Stop();
        Start();
    }

    public void RefreshNow()
    {
        _ = Task.Run(() => Tick(CancellationToken.None, force: true));
    }

    private async Task Tick(CancellationToken ct, bool force = false)
    {
        if (_store.IsPaused && !force) return;
        // A forced tick may race the periodic loop — never run two interleaved.
        if (_tickInFlight) return;
        _tickInFlight = true;
        try
        {
            var settings = _store.Settings;

            // UsageStore mutations synchronously raise PropertyChanged, which
            // ViewModels handle by constructing Avalonia brushes/colors —
            // Avalonia objects can only be created on the UI thread, so every
            // store mutation must be dispatched there even though this whole
            // method runs on a background Task (the actual file/network I/O
            // below stays off-thread; only the store touches are marshaled).
            await Dispatcher.UIThread.InvokeAsync(() =>
            {
                foreach (var provider in _providers) _store.MarkRefreshing(provider.Id);
            });

            var tasks = _providers.Select(async provider =>
            {
                try
                {
                    var snapshot = await provider.FetchSnapshotAsync(settings, ct);
                    return (provider.Id, Snapshot: (UsageSnapshot?)snapshot, Error: (string?)null);
                }
                catch (Exception ex)
                {
                    return (provider.Id, Snapshot: (UsageSnapshot?)null, Error: ex.Message);
                }
            });
            var results = await Task.WhenAll(tasks);

            await Dispatcher.UIThread.InvokeAsync(() =>
            {
                foreach (var (id, snapshot, error) in results)
                {
                    if (snapshot is not null)
                    {
                        _store.Apply(snapshot);
                        Log.Refresh.LogDebug("refreshed {0}: {1}", id, snapshot.Health);
                    }
                    else
                    {
                        _store.ApplyFailure(id, error ?? "unknown error");
                        Log.Refresh.LogError("refresh failed {0}: {1}", id, error ?? "unknown error");
                    }
                }
            });
            _snapshotStore.Save(_store.Snapshots);
        }
        catch (Exception ex)
        {
            // A single bad tick (e.g. a transient file lock) must never kill
            // the loop silently — `_ = Task.Run(...)` in Start() discards the
            // task, so an uncaught exception here would stop all future
            // refreshes forever with no visible error until the app restarts.
            Log.Refresh.LogError("tick failed: {0}", ex);
        }
        finally
        {
            _tickInFlight = false;
        }
    }
}
